server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala

```diff
@@ -250,31 +253,67 @@ class EmailSubmissionSetMethod @Inject()(serializer: EmailSubmissionSetSerialize
+      delay <- SMono.fromTry(retrieveDelay(parameters))
+      _ <- SMono.fromTry(validateDelay(delay))
+
       mail = {
         val mailImpl = MailImpl.builder()
           .name(submissionId.value)
-          .addRecipients(envelope.rcptTo.map(_.email).asJava)
+          .addRecipients(envelope.rcptTo.filter(m => m.parameters.isEmpty || !(m.parameters.get.contains(ParameterName.holdFor) || m.parameters.get.contains(ParameterName.holdUntil))).map(_.email).asJava)
           .sender(envelope.mailFrom.email)
           .addAttribute(new Attribute(MAIL_METADATA_USERNAME_ATTRIBUTE, AttributeValue.of(mailboxSession.getUser.asString())))
           .build()
         mailImpl.setMessageNoCopy(message)
         mailImpl
       }
-      _ <- SMono(queue.enqueueReactive(mail))
+
+      _ <- SMono(queue.enqueueReactive(mail, delay))
         .`then`(SMono.fromCallable(() => LifecycleUtil.dispose(mail)).subscribeOn(Schedulers.boundedElastic()))
         .`then`(SMono.just(submissionId))
+      sendAt = UTCDate(ZonedDateTime.now(clock).plus(delay))
     } yield {
-      EmailSubmissionCreationResponse(submissionId) -> request.emailId
+      EmailSubmissionCreationResponse(submissionId, sendAt) -> request.emailId
+    }
+
+  private def retrieveDelay(mailParameters: Option[Map[ParameterName, Option[ParameterValue]]]): Try[Duration] = mailParameters match {
+      case None => Success(Duration.ZERO)
+      case Some(aMap) if aMap.contains(ParameterName.holdFor) => Try(Duration.ofSeconds(aMap(ParameterName.holdFor).getOrElse(ParameterValue("0")).value.toLong))
+      case Some(aMap) if aMap.contains(ParameterName.holdUntil) && aMap(ParameterName.holdUntil).nonEmpty => Try(Duration.between(LocalDateTime.now(clock), LocalDateTime.parse(aMap(ParameterName.holdUntil).get.value, formatter)))
+      case Some(aMap) if aMap.contains(ParameterName.holdUntil) && aMap(ParameterName.holdUntil).isEmpty => Success(Duration.ZERO)
+      case _ => Success(Duration.ZERO)
     }
 
+  def validateDelay(delay: Duration): Try[Duration] =
+    if (delay.getSeconds >= 0 && delay.getSeconds <= SubmissionCapabilityFactory.maximumDelays.getSeconds) {
+      Success(delay)
+    } else {
+      Failure(new IllegalArgumentException("Invalid Arguments"))
+    }
+
+  def validateFromParameters(mailParameters: Option[Map[ParameterName, Option[ParameterValue]]]): Try[Option[Map[ParameterName, Option[ParameterValue]]]] = {
+    val keySet: Set[ParameterName] = mailParameters.getOrElse(Map()).keySet
+    val invalidEntries = keySet -- Set(ParameterName.holdFor, ParameterName.holdUntil)
+    if (invalidEntries.isEmpty) {
+      if (keySet.contains(ParameterName.holdFor) && keySet.contains(ParameterName.holdUntil)) {
+        Failure(new IllegalArgumentException("Can't specify holdFor and holdUntil simultaneously"))
+      } else {
+        Success(mailParameters)
+      }
+    } else {
+      Failure(new IllegalArgumentException("Unsupported parameterName"))
+    }
+  }
```

Comment:
This does not follow O (open for extension but closed for modification) in SOLID IMO.

With this code, it is hard to extend more parameters in the future, and it requires a lot of modifications when adding them (more nested if else IMO).